Added a Translate function

ctSkennerton opened the pull request:
new translate function takes a nucleotide string and returns the amino acid equivalent.  Can optionally specify any of the translation tables listed from NCBI.

Commit:
initial commit for github

Commit:
should have been in initial list

ctSkennerton committed:
Added in a translate function

ctSkennerton committed:
Fixed some formatting in README.md

ctSkennerton committed:
Fix off-by-1 error when determining the translation table requested

ctSkennerton committed:
Fix switch statement fall through

Commit:
awk: Use random(3) instead of rand(3)

While none of them is considered even near to cryptographic
level, random(3) is a better random generator than rand(3).

Use random(3) for awk as is done in other systems.

Thanks to Chenguang Li for discussing this in the lists
and submitting the patch upstream.

PR:		193147
MFC after:	5 weeks

git-svn-id: svn+ssh://svn.freebsd.org/base/head@271879 ccf9f872-aa2e-dd11-9fc8-001c23d0bc1f

mikheyev commented:
This is an essential function, please add it to the repo. I already use it. :+1: :100:

Commit:
MFV r300961: one-true-awk: replace 0 with NULL for pointers

Also remove a redundant semicolon.

Also had to rebase on upstream pull.

git-svn-id: svn+ssh://svn.freebsd.org/base/head@301289 ccf9f872-aa2e-dd11-9fc8-001c23d0bc1f

ConradStack committed:
-added minqual, maxqual functions

ConradStack committed:
fixed .o issue

ConradStack committed:
ignore macosx system files

ConradStack committed:
minor change to maxqual

Commit:
awktest.tar added; .ar files work poorly

Commit:
MFV r315425: one-true-awk: have calloc(3) do the multiplication.

MFC after:	3 days

git-svn-id: svn+ssh://svn.freebsd.org/base/head@315426 ccf9f872-aa2e-dd11-9fc8-001c23d0bc1f

ConradStack committed:
fixed average quality score calculation

Commit:
set baseline so Arnold can send pull request

Commit:
update awktest.tar with proper beebe,lilly tests

Commit:
FIXES, makefile, REGRESS to smooth out base for Arnold

arnoldrobbins committed:
Fix multiple long-standing bugs, improve test suite.

onetrueawk committed:
Merge pull request #8 from arnoldrobbins/master

Fix multiple long-standing bugs, improve test suite.

Commit:
added Arnold's fixes, updates to awktest.tar

Commit:
disallow $ in printf formats

Commit:
forgot to update version string

atsampson committed:
Fix comment syntax for pmake.

GNU make is happy with leading tabs before a comment, but pmake says
"Unassociated shell command".

atsampson committed:
Fix ytab.[ch] dependencies to allow parallel make.

The old ytab.o rule produced ytab.c and ytab.h too, but this wasn't
visible as dependencies, so make -j would fail. Make this explicit, and
use the implicit .c.o rule rather than writing out the $(CC) command.

onetrueawk committed:
Merge pull request #9 from atsampson/master

makefile fixes: allow parallel make and work with pmake

iamleot committed:
avoid undefined behaviour when using ctype(3) functions in relex()

Because NCHARS is (256+3) cc->cc_func(i) was called with 256, 257
and 258 as argument leading to possible undefined behaviour (at
least on NetBSD with non-C locale (e.g. `en_US.UTF-8') this led to
only honoring one `[:...:]' character class in bracket expressions).

Fix #11

arnoldrobbins committed:
Fix REGRESS so all tests run on Mac OS X.

melloc committed:
Check for format character precision argument before using it

melloc committed:
Protect against overflowing during OFMT/CONVFMT conversions

melloc committed:
Fix issues with assigning during concatenation

melloc committed:
Fix issues with numeric SUBSEP and large SUBSEP values

melloc committed:
Fix calling split() with a third argument that lives in the target array

melloc committed:
Handle numeric FS, RS, OFS, and ORS values

melloc committed:
Improve error reporting messages

melloc committed:
Rebuild fields when NF is assigned to itself

melloc committed:
Print an error message for negative NF values

melloc committed:
Flags that begin with "--" should not be treated as end of args

melloc committed:
Update field-splitting behaviour to match POSIX definition

Commit:
fix maketab non-bug

melloc committed:
Fix regular expressions containing [[:cntrl:]]

melloc committed:
Fix initial "fields" buffer size

ctSkennerton committed:
Merge branch 'master' of https://github.com/ConradStack/bioawk

ctSkennerton committed:
Merge branch 'master' of https://github.com/lh3/bioawk

ctSkennerton committed:
update man file

ctSkennerton committed:
forgot about bed format

junghans committed:
fixed parallel build

Signed-off-by: Christoph Junghans <junghans@gentoo.org>

Commit:
add in the lshift, rshift and compl functions

these functions round out the bitwise operators

onetrueawk committed:
Merge pull request #13 from arnoldrobbins/master

Fix REGRESS so all tests run on Mac OS X.

onetrueawk committed:
Merge pull request #14 from melloc/missing-precision

Check for format character precision argument before using it

onetrueawk committed:
Merge branch 'master' into assign-expr

onetrueawk committed:
Merge branch 'master' into fmt-overflow

onetrueawk committed:
Merge pull request #15 from melloc/fmt-overflow

Protect against overflowing during OFMT/CONVFMT conversions

onetrueawk committed:
Merge branch 'master' into subsep

onetrueawk committed:
Merge pull request #18 from melloc/error-reporting

Improve error reporting messages

onetrueawk committed:
Merge branch 'master' into split-fs-from-array

onetrueawk committed:
Merge branch 'master' into nf-self-assign

onetrueawk committed:
Merge pull request #21 from melloc/end-of-args

Flags that begin with "--" should not be treated as end of args

onetrueawk committed:
Merge pull request #23 from melloc/cntrl-class

Fix regular expressions containing [[:cntrl:]]

onetrueawk committed:
Merge pull request #24 from melloc/fields-init

Fix initial "fields" buffer size

onetrueawk committed:
Merge branch 'master' into assign-expr

onetrueawk committed:
Merge pull request #16 from melloc/assign-expr

Fix issues with assigning during concatenation

onetrueawk committed:
Merge branch 'master' into subsep

onetrueawk committed:
Merge pull request #17 from melloc/subsep

Fix issues with numeric separator values

onetrueawk committed:
Merge branch 'master' into split-fs-from-array

onetrueawk committed:
Merge pull request #19 from melloc/split-fs-from-array

Fix calling split() with a third argument that lives in the target array

onetrueawk committed:
Merge branch 'master' into nf-self-assign

onetrueawk committed:
Merge pull request #20 from melloc/nf-self-assign

Fix issues with NF self-assignment and setting NF to negative values

onetrueawk committed:
Merge pull request #27 from junghans/fix_parallel_build

fixed parallel build

McDutchie committed:
backport ERE interval/repetition expressions from Apple awk-24

The lack of POSIX interval expressions[*] (a.k.a. bounds, a.k.a.
repetition expressions) in regular expressions is listed under BUGS
in 'awk.1'. Apple's version of onetrueawk has supported these since
at least 2009, judging by the date stamp on their src/b.c in:
https://opensource.apple.com/tarballs/awk/awk-24.tar.gz

A bug report prompted NetBSD to swiftly integrate this code into
their awk. This commit is based on that NetBSD diff.
http://gnats.netbsd.org/53885
https://github.com/NetBSD/src/commit/f3e4c4ca1dfcdd939a2e33ebfe708f01e25b3bae

b.c:
- Backport POSIX-standard interval expressions support in regular
  expressions via NetBSD from Apple awk-24 (20070501).

main.c:
- Bump version ID.

FIXES:
- Add note and credit for this feature.

awk.1: section BUGS:
- Remove line saying interval expressions are not supported.

_________
[*] http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04_06

arnoldrobbins committed:
Make getline POSIX compliant w.r.t. numeric strings.

arnoldrobbins committed:
Update FIXES.

arnoldrobbins committed:
Simplify getline-numeric input and results.

arnoldrobbins committed:
Update version string in main.c.

iamleot committed:
Merge branch 'master' into cc_func-avoid-undefined-behaviour

enh-google committed:
maketab: support build systems with read-only source.

If your generated files are considered outputs that live elsewhere, you
need a way to tell maketab where ytab.h actually is.

Specifically, I'm trying to avoid checking in generated files in
Android's AOSP tree's copy of one-true-awk.

enh-google committed:
Add .gitignore.

McDutchie committed:
repeat(): add FATAL calls for errors that should be impossible

onetrueawk committed:
Merge pull request #12 from iamleot/cc_func-avoid-undefined-behaviour

Avoid undefined behaviour when using ctype(3) functions in relex() (possible fix for #11)

onetrueawk committed:
Merge pull request #31 from arnoldrobbins/master

Make getline handle numeric strings, and update FIXES

onetrueawk committed:
Merge pull request #32 from enh/master

maketab: support build systems with read-only source.

McDutchie committed:
Merge remote-tracking branch 'upstream/master' into interval-expr

main.c: bump version to 20190305

arnoldrobbins committed:
Merge pull request #30 from McDutchie/interval-expr

backport ERE interval/repetition expressions from Apple awk-24

arnoldrobbins committed:
Sync FIXES with GitHub repo activity.

Commit:
Fix cross-build (#34)

* Fix cross-build

Fixes in make file to support cross-build because maketab is a host tool and should be compiled with host compiler

* some

arnoldrobbins committed:
Simplify cross-compiling change. Update FIXES.

arnoldrobbins committed:
Update awktest.tar(p.50) to modern sort options.

ctSkennerton committed:
Merge branch 'master' of https://github.com/onetrueawk/awk

ctSkennerton committed:
Merge branch 'master' of github.com:ctSkennerton/bioawk

ctSkennerton committed:
change the name of the man file

ctSkennerton committed:
update makefile

ctSkennerton committed:
change case of makefile

ctSkennerton committed:
minor updates to man file

arnoldrobbins committed:
Fix Issue #38 - don't require non-= after = in cmd line assignment.

bsdimp committed:
Bring in fix from FreeBSD:

| r323964 | imp | 2017-09-23 23:04:02 -0600 (Sat, 23 Sep 2017) | 6 lines
|
| Fix %c for floating values that become 0 when coerced to int.
|
| Obtained from: OpenBSD run.c 1.36 (From Jeremy Devenport)
| Sponsored by: Netflix
| Differential Revision: https://reviews.freebsd.org/D12379

bsdimp committed:
Apply r323963 from FreeBSD (pulling in the fix from OpenBSD)

| Fix uninitialized variable
|
| echo | awk 'BEGIN {i=$1; print i}' prints a boatload of stack
| garbage. NUL terminate the memory returned from malloc to prevent it.
|
| Obtained from: OpenBSD run.c 1.40
| Sponsored by: Netflix
| Differential Revision: https://reviews.freebsd.org/D12379

bsdimp committed:
Apply the following from FreeBSD / OpenBSD:

323965 | imp | 2017-09-23 23:04:06 -0600 (Sat, 23 Sep 2017) | 8 lines

Don't display empty error context.

Context extraction didn't handle this case and showed uninitialized memory.

Obtained from: OpenBSD lib.c 1.21
Sponsored by: Netflix
Differential Revision: https://reviews.freebsd.org/D12379

arnoldrobbins committed:
Allow unmatched right paren in regexes. Fixes Issue #40.

arnoldrobbins committed:
Minor fixes in the man page.

ctSkennerton committed:
fixes to the man page

ctSkennerton committed:
Merge branch 'master' of https://github.com/onetrueawk/awk

ctSkennerton committed:
expose tags field from sam files

ctSkennerton committed:
add in a sam tag filtering example

ctSkennerton committed:
update link to one true awk

melloc committed:
Disallow deleting SYMTAB and its elements (#43)

arnoldrobbins committed:
Extract testdir.

arnoldrobbins committed:
Remove awktest.tar.

arnoldrobbins committed:
Update makefile, mainly to clean up testdir.

arnoldrobbins committed:
Update FIXES for test suite changes.

ctSkennerton committed:
Merge branch 'master' of https://github.com/onetrueawk/awk

arnoldrobbins committed:
Merge remote-tracking branch 'melloc/posix-fs' into save-fs-fixes

arnoldrobbins committed:
Update T.split to match code changes.

arnoldrobbins committed:
Revise testdir/T.split per PR #42.

arnoldrobbins committed:
Update FIXES and version in main.c.

arnoldrobbins committed:
Update FIXES and main.c version.

Commit:
Merge branch 'master' of https://github.com/onetrueawk/awk

McDutchie committed:
Add support for "\a" and "\v" to regex and command line args (#44)

Support POSIX-specified C-style escape sequences "\a" (alarm)
and "\v" (vertical tab) in command line arguments and regular
expressions, further to the support for them in strings added on
Apr 9, 1989. These now no longer match as literal "a" and "v"
characters (as they don't on gawk and mawk).

IOW, lex.c already supported these (lines 390-391 as of 4e343460);
the support needed to be added to b.c and tran.c.

Relevant POSIX reference:
http://pubs.opengroup.org/onlinepubs/9699919799/utilities/awk.html#tag_20_06_13_04

arnoldrobbins committed:
Improve cleanup of test directory upon 'make clean'.

arnoldrobbins committed:
Remove trailing whitespace on lines in all files.

arnoldrobbins committed:
Grammar optimization from NetBSD: Two adjacent string constants are merged.

ctSkennerton committed:
Merge branch 'master' of https://github.com/onetrueawk/awk

ctSkennerton committed:
ignore the bioawk binary

sevan committed:
Drop Mac OS 9 reference (#46)

"buildmac" file is no longer supplied.

ctSkennerton committed:
fix bug where the OFS wasn't being honoured for fastx

ctSkennerton committed:
add in test rules for bio extensions

arichardson committed:
Fix out-of-bounds access in gototab array for caret character (#47)

When matching a caret, the expression	`f->gototab[s][c] = f->curstat;` in
cgoto() will index the 2D-array gototab with [s][261]. However, gototab
is declared as being of size [NSTATES][NCHARS], so [32][259]. Therefore,
this assignment will write to the state for character 0x1.
I'm not sure how to create a regression test for this, but increasing the
array size to HAT+1 values fixes the error and the tests still pass.

I found this issue while running awk on a CHERI system with sub-object
protection enabled. On x86, this can be reproduced by compiling awk
with -fsanitize=undefined.

arichardson committed:
Avoid creating an out-of-bounds pointer in word() (#48)

As it is never dereferenced in the n == -1 case it shouldn't cause any
problems. However, UBSAN complains about this, so it is required to run
the tests when compiling with -fsanitize=undefined.

arnoldrobbins committed:
Update FIXES and version date in main.c.

ctSkennerton committed:
added in test case for issue #2

ctSkennerton committed:
fix issue #2.

The field separator wasn't being honoured because upstream changes
had modified the use of FS in the code. The function `savefs` now
needs to be called after a record is read in.

ctSkennerton committed:
add in versioning based on git tags

ctSkennerton committed:
Merge branch 'master' of https://github.com/onetrueawk/awk

ctSkennerton committed:
Improvements to the man file
